src/model/DaylightingDeviceTubular.hpp

```diff
@@ -0,0 +1,158 @@
+*  STATES DEPARTMENT OF ENERGY, NOR ANY OF THEIR EMPLOYEES, BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+*  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+*  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+*  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***********************************************************************************************************************/
+
+#ifndef MODEL_DAYLIGHTINGDEVICETUBULAR_HPP
+#define MODEL_DAYLIGHTINGDEVICETUBULAR_HPP
+
+#include "ModelAPI.hpp"
+#include "ModelObject.hpp"
+
+#include "ThermalZone.hpp"
+#include "ThermalZone_Impl.hpp"
+
+namespace openstudio {
+namespace model {
+
+  class SubSurface;
+  class ConstructionBase;
+  class ThermalZone;
+
+  namespace detail {
+
+    class DaylightingDeviceTubular_Impl;
+
+  }  // namespace detail
+
+  /** This class implements a transition zone */
+  class MODEL_API TransitionZone
+  {
+   public:
+    TransitionZone(const ThermalZone& zone, double length);
+
+    ThermalZone thermalZone() const;
+    double length() const;
+
+   private:
+    ThermalZone m_zone;
+    double m_length;
+    REGISTER_LOGGER("openstudio.model.TransitionZone");
+  };
+
+  // Overload operator<<
+  MODEL_API std::ostream& operator<<(std::ostream& out, const openstudio::model::TransitionZone& transitionZone);
+
+  /** DaylightingDeviceTubular is a ModelObject that wraps the OpenStudio IDD object 'OS_DaylightingDevice_Tubular'. */
+  class MODEL_API DaylightingDeviceTubular : public ModelObject
+  {
+   public:
+    /** @name Constructors and Destructors */
+    //@{
+
+    explicit DaylightingDeviceTubular(const SubSurface& dome, const SubSurface& diffuser, const ConstructionBase& construction, double diameter,
+                                      double totalLength, double effectiveThermalResistance);
+
+    explicit DaylightingDeviceTubular(const SubSurface& dome, const SubSurface& diffuser, const ConstructionBase& construction);
+
+    virtual ~DaylightingDeviceTubular() {}
+
+    //@}
+    /** @name Static Methods */
+    //@{
+
+    static IddObjectType iddObjectType();
+
+    //@}
+    /** @name Getters */
+    //@{
+
+    SubSurface subSurfaceDome() const;
+
+    SubSurface subSurfaceDiffuser() const;
+
+    ConstructionBase construction() const;
+
+    double diameter() const;
+
+    double totalLength() const;
+
+    double effectiveThermalResistance() const;
+
+    std::vector<TransitionZone> transitionZones() const;
+
+    unsigned int numberofTransitionZones() const;
+
+    //@}
+    /** @name Setters */
+    //@{
+
+    bool setConstruction(const ConstructionBase& construction);
```

Comment:
Yes, intentional. I'm following the existing `DaylightingDeviceShelf` as a guide.

Reply:
It's kinda annoying if you want to clone the `DaylightingDeviceTubular` no? (to have the numeric fields etc all filed out, before assigning it to another surface which could be in the same zone)

Reply:
Are you proposing having subsurface setters on all DaylightingDevice objects? e.g., including DaylightingDeviceShelf which currently does not have it?

Reply:
I guess that's what I was suggesting yes. 

I'm probably overthinking this though: it's not objects that are heavily used, and no one has complained about cloning shelves and they have been around for a while.

That being said, cloning a shelf had less interest because it doesn't really have any fields that shouldn't be unique (it references inside/outside shelf surfaces and a view factor only)

Anyways, let's leave the code as is currently